server/bleep/src/webserver/answer.rs

```diff
@@ -94,34 +98,141 @@ pub async fn handle(
         super::error(ErrorKind::Internal, "semantic search returned no snippets");
     }
 
-    let res = reqwest::Client::new()
-        .post(format!("{}/q", app.config.answer_api_base))
-        .json(&api::Request {
-            query: params.q,
-            user_id: params.user_id,
-            snippets: snippets.clone(),
-        })
-        .send()
+    let answer_api_client = AnswerAPIClient::new(
+        &format!("{}/q", app.config.answer_api_base),
+        &target,
+        &params.user_id,
+    );
+
+    let res1 = answer_api_client
+        .stage_1(&snippets)
+        .await
+        .map_err(|e| super::error(ErrorKind::Internal, e.to_string()))?
+        .index;
+
+    let processed_snippet = &snippets[res1];
+
+    let res2 = answer_api_client
+        .stage_2(processed_snippet)
         .await
-        .map_err(|e| {
-            super::error(
-                ErrorKind::Internal,
-                format!("failed to make request to answer API: {}", e),
-            )
-        })?;
-
-    let selection: api::Response = res.json().await.map_err(|e| {
-        super::error(
-            ErrorKind::Internal,
-            format!("answer API was not able to create a valid result: {}", e),
-        )
-    })?;
-
-    let selected_snippet = snippets.remove(selection.data.index as usize);
+        .map_err(|e| super::error(ErrorKind::Internal, e.to_string()))?
+        .answer;
+
+    // reorder snippets
+    let selected_snippet = snippets.remove(res1);
     snippets.insert(0, selected_snippet);
```

Comment:
Is it important that the other snippets keep their order? If not, we could use `swap` here. Otherwise, I believe that `snippets[..(res1 + 1)].rotate_right(1)` should do the same thing with less overhead.

Reply:
the order of the other snippets is not important, we can absolutely use `swap` here.